lib/bounce_email.rb

```diff
@@ -237,5 +239,27 @@ def extract_original_message_after_delimiter(mail, delimiter_index)
       message = mail.body.to_s.split(delimiter).last
       message.split(INLINE_MESSAGE_END_DELIMITER).first.strip if message.match(INLINE_MESSAGE_END_DELIMITER)
     end
+
+    def extract_and_assign_fields_from_original_mail(mail)
+      if mail.message_id.nil?
+        mail.add_message_id extract_field_from(mail, 'Message-ID:')
+      end
+
+      mail.from ||= extract_field_from(mail, 'From:')
+      mail.to ||= extract_field_from(mail, 'To:')
+      mail.subject ||= extract_field_from(mail, 'Subject:')
+
+      mail
+    end
+
+    def extract_field_from(mail, field_name)
+      lines = original_mail_body_lines(mail)
+      field = lines.detect { |line| line.match field_name }
+      field.split(':', 2).last.strip if field
+    end
+
+    def original_mail_body_lines(mail)
+      @original_mail_body_lines ||= mail.body.to_s.split(/\R+/)
```

Comment:
Neat! I didn't know about `\R`. Where did you find documentation about it? I was unable to locate anything in the docs for the `Regexp` class.

Sadly, my tests show that `\R` works only in Ruby 2.0 and newer, but not in 1.8.7 and 1.9.x. We should probably just use `/[\r\n]+/`.